src/subcommand/init/shell/posix.rs

```diff
@@ -37,12 +37,13 @@ _z_cd() {{
             return 1
         fi
     else
-        result="$(zoxide query "$@")" || return "$?"
-        if [ -d "$result" ]; then
-            _z_cd "$result" || return "$?"
-        elif [ -n "$result" ]; then
-            echo "$result"
+        _zoxide_result="$(zoxide query "$@")" || return "$?"
+        if [ -d "$_zoxide_result" ]; then
+            _z_cd "$_zoxide_result" || (unset _zoxide_result && return "$?")
```

Comment:
Calling `unset` here will reset the exit code that we're returning (`$?`). I think it should be safe to not use `unset`, since the variable is "namespaced" under `_zoxide_`.